core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -970,8 +970,39 @@ class Partition(val topicPartition: TopicPartition,
       tryCompleteDelayedRequests()
   }
 
+  /**
+   * We should only shrink the ISR if the ISR set is >= minISR after the shrinking.
+   * When ISR set needs to be dropped under MinISR, a leadership switch would be triggered.
+   */
   private def needsShrinkIsr(): Boolean = {
-    leaderLogIfLocal.exists { _ => getOutOfSyncReplicas(replicaLagTimeMaxMs).nonEmpty }
+    leaderLogIfLocal.exists { log =>
+      val outOfSyncReplicaIds = getOutOfSyncReplicas(replicaLagTimeMaxMs)
+      outOfSyncReplicaIds.nonEmpty && (inSyncReplicaIds -- outOfSyncReplicaIds).size >= log.config.minInSyncReplicas
+    }
+  }
+
+  def maybeTransferToNewLeader(): Unit = {
+    val recommendedNewLeader = inReadLock(leaderIsrUpdateLock) {
+      needsLeadershipTransfer()
+    }
+
+    recommendedNewLeader match {
+      case Some(newLeader) =>
+        transferToNewLeader(newLeader)
+      case None => // do nothing
+    }
+  }
+  /**
+   * We should transfer the leadership if the ISR needs to be dropped to be below MinISR
+   * @return the recommended new leader for the partition, which is the first broker in the ISR
+   */
+  private def needsLeadershipTransfer(): Option[Int] = {
+    leaderLogIfLocal.flatMap { log =>
+      val outOfSyncReplicaIds = getOutOfSyncReplicas(replicaLagTimeMaxMs)
+      if ((inSyncReplicaIds -- outOfSyncReplicaIds).size < log.config.minInSyncReplicas) {
+        inSyncReplicaIds.find(_ != localBrokerId)
```

Comment:
I thought more about this. Suppose we have the following condition with 3 replicas in the ISR:
leader A has 5 messages (offsets 1-5);
follower B has 4 messages (offsets 1-4);
follower C has 3 messages (offsets 1-3).

One might think that transferring the new leader to B is the preferable option. However, because the high watermark is the lowest end offset of the ISR, i.e. determined by follower C in this case,
regardless of whether we transfer the leader to B or C, the produce request for msg at offset 4 and 5 will receive a "NotLeaderForPartition" error and retry the produce requests.
Thus transferring the leadership to B does not actually help.

Besides this line of thinking, I'm not sure how else the handoff algorithm can be smarter.